src/main/assisted-installer-controller/assisted_installer_main.go

```diff
@@ -304,3 +284,63 @@ func didInstallationFinish(clusterStatus string, log logrus.FieldLogger, status
 	}
 	return false
 }
+
+func getClusterAndDetermineTaintRemoval(ctx context.Context, assistedController assistedinstallercontroller.Controller,
+	ic inventory_client.InventoryClient, kc k8s_client.K8SClient, logger logrus.FieldLogger, invoker string) (*models.Cluster, bool) {
+	var cluster *models.Cluster
+	removeUninitializedTaint := false
+	if invoker == common.InvokerAgent {
+		if Options.ControllerConfig.HighAvailabilityMode == models.ClusterHighAvailabilityModeNone {
+			// When the agent-based installer installs a SNO cluster, assisted-service
+			// will never be reachable because it will not be running after the boostrap
+			// node reboots to become the SNO cluster. SetReadyState will never be
+			// able to get a connection to assisted-service in this circumstance.
+			//
+			// Instead of exiting with panic with "invalid memory address or nil pointer
+			// dereference" and having the controller restart because cluster is nil,
+			// log warning and exit 0. Otherwise the controller will keep restarting
+			// leaving the controller in a running state, even through the cluster
+			// has finished install.
+			logger.Warnf("cluster is SNO and invoker = %v, skipping assisted-installer-controller", invoker)
+			os.Exit(0)
+		}
+		// With the agent-based installer, assisted-service runs on the bootstrap node.
+		// We do not need to wait for an extended amount of time to reach assisted-service
+		// because it should be running on the local network.
+		cluster = assistedController.SetReadyState(5 * time.Minute)
```

Comment:
I don't think this is the right function to use to say the truth, seems not relevant to way for ready state here, you need cluster from the service and not to set ready even on k8s

Reply:
What would you suggest to be a better function to use here?

In the majority of the time, ABI should be able to retrieve the cluster from assisted-service. For ABI, if assisted-service isn't unavailable, we just want to fail more quickly (in 5 minutes instead of 70 minutes).

In the customer case for the bug, assisted-service is unavailable, but k8s is available. In this case, we use the data in k8s to determine whether the taint should be removed and try to move forward and allow CSR approval to run so that the remaining nodes can join the cluster and have the cluster installation succeed.

Reply:
maybe lets create new one that just waits? i mean you can fail here cause of 2 different things. 
polling service and polling k8s, maybe just try to get cluster from assisted?

Reply:
I think it is fine they way it is actually. For the common case, it should check that both assisted-service and kube-apiserver are ready.

If assisted-service is ready but the k8s poll fails, then it should wait until it succeeds, otherwise the next step where it tries the determine the platform type through the ConfigMap would fail.

If assisted-service is not ready, then the check on kube-apiserver would be skipped. I'm not sure if it is worth adding a check or wait here for kube-apiserver. If it is not ready, then the controller will fail and restart. 

Does that sound ok to you?